addons/mail_client_extension/controllers/main.py

```diff
@@ -0,0 +1,323 @@
+                'enrichment_info': None
+            }
+            company = self._find_existing_company(sender_domain)
+            if not company:  # create and enrich company
+                company, enrichment_info = self._create_company_from_iap(sender_domain)
+                response['enrichment_info'] = enrichment_info
+            response['partner']['company'] = self._get_company_dict(company)
+
+        return response
+
+    @http.route('/mail_client_extension/partner/create', type="json", auth="outlook", csrf=False, cors="*")
+    def res_partner_create(self, email, name, company, **kwargs):
+        # TODO search the company again instead of relying on the one provided here?
+
+        # Create the partner if needed.
+        partner_info = {
+            'name': name,
+            'email': email,
+        }
+        if company > -1:
+            partner_info['parent_id'] = company
+        partner = request.env['res.partner'].with_user(request.env.user).create(partner_info)
+
+        response = {
+            'authorized': True,
+            'created': True,
+            'id': partner.id
+        }
+        return response
+
+    @http.route('/mail_client_extension/log_single_mail_content', type="json", auth="outlook", csrf=False, cors="*")
+    def log_single_mail_content(self, lead, message, **kw):
+        response = {
+            'authorized': True,
+            'success': True
+        }
+        crm_lead = request.env['crm.lead'].with_user(request.env.user).browse(lead)
+        try:
+            crm_lead._message_log(body=message, author_id=int(request.env.user))
+        except:# Exception as e:
+            response['success'] = False
+
+        return response
+
+    @http.route('/mail_client_extension/lead/get_by_partner_id', type="json", auth="outlook", csrf=False, cors="*")
+    def crm_lead_get_by_partner_id(self, partner, limit, offset, **kwargs):
+        partner_leads = request.env['crm.lead'].with_user(request.env.user).search([('partner_id', '=', partner)], offset=offset, limit=limit)
+        leads = []
+        for lead in partner_leads:
+            leads.append({
+                'id': lead.id,
+                'name': lead.name,
+                'expected_revenue': str(lead.expected_revenue),
+                'planned_revenue': str(lead.planned_revenue),
+                'currency_symbol': lead.company_currency.symbol
+            })
+
+        return {
+            'authorized': True,
+            'leads': leads
+        }
+
+    @http.route('/mail_client_extension/lead/redirect_form_view', type='http', auth='user', methods=['GET'])
+    def crm_lead_redirect_form_view(self, partner_id):
+        # localhost:8069/crm/lead/redirect_form_view?partner_id=42
+        server_action = http.request.env.ref("mail_client_extension.lead_creation_prefilled_action")
+        return werkzeug.utils.redirect('/web?&#action=%s&model=crm.lead&view_type=form&partner_id=%s' % (server_action.id, partner_id))
```

Comment:
`parther_id` is user-provided, we should probably validate that it's an integer (e.g. either `int(partner_id)` or check that `partner_id.isdigit()` and generate some sort of error otherwise) before just merging it into a URL as-is.

Also I think it'd already noted that the `?&` should be removed.